--- lstrlib.c
+++ lstrlib.c
@@ -19,6 +19,8 @@
 #include "lauxlib.h"
 #include "lualib.h"
 
+#include "ldo.h"
+
 
 /* macro to `unsign' a character */
 #define uchar(c)        ((unsigned char)(c))
@@ -363,6 +365,11 @@
 
 
 static const char *match (MatchState *ms, const char *s, const char *p) {
+  /* each level of a pattern's match is a C call deeper */
+  switch (luai_cstack()) {
+    case 1: luaL_error(ms->L, "pattern too complex");
+    case 2: luaD_throw(ms->L, LUA_ERRERR);
+  }
   init: /* using goto's to optimize tail recursion */
   switch (*p) {
     case '(': {  /* start capture */
